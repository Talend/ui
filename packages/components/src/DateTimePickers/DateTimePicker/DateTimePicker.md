# DateTimePicker

This component display a datetime picker

## props

- **selection** : The data used to control the current selection
    - object format with :
        - _Date_ prop as a Date object
        - _time_ prop as a number of seconds since the beginning of the date
- **onSubmit** : Trigger when defined datetime picked is submit
    - return in first arg an object with the same format as _selection_ prop
