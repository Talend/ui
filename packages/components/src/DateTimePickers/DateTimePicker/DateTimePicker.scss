.container {
	min-height: 25rem;

	button:focus {
		outline: 1px solid black;
	}
}
