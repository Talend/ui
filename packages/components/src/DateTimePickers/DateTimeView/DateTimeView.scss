
@import '../common';

.action-left, .action-right  {
    @include nav-action-icon(12px);
}

.action-middle {
    @include main;
    @include nav-action-default;
    font-size: 16px;
}

.body {
    display: flex;

    .date {
        flex: 3;
    }

    .time {
        flex: 1;
        border-left: $border-default;
    }
}
