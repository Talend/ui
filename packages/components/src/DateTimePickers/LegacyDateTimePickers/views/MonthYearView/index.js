import { withTranslation } from 'react-i18next';
import MonthYearView from './MonthYearView.component';
import I18N_DOMAIN_COMPONENTS from '../../../../constants';

export default withTranslation(I18N_DOMAIN_COMPONENTS)(MonthYearView);
