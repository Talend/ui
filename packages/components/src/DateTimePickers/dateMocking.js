let OriginalDate;

export function mockDate(mockingDate) {
	if (OriginalDate === undefined) {
		OriginalDate = Date;
	}

	global.Date = jest.fn((...args) => {
		if (args.length === 0) {
			return mockingDate;
		}
		return new OriginalDate(...args);
	});
}


export function restoreDate() {
	global.Date = OriginalDate;
	OriginalDate = undefined;
}
