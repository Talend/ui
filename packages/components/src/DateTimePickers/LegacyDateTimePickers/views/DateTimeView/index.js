import { withTranslation } from 'react-i18next';
import DateTimeView from './DateTimeView.component';
import I18N_DOMAIN_COMPONENTS from '../../../../constants';

export default withTranslation(I18N_DOMAIN_COMPONENTS)(DateTimeView);
