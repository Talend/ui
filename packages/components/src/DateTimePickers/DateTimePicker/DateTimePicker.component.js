import React from 'react';
import PropTypes from 'prop-types';
import getMonth from 'date-fns/get_month';
import getYear from 'date-fns/get_year';

import theme from './DateTimePicker.scss';
import DateTimeView from '../views/DateTimeView';
import MonthYearView from '../views/MonthYearView';

class DateTimePicker extends React.Component {
	constructor(props) {
		super(props);

		// eslint-disable-next-line
		console.warn(
			"UNSTABLE WARNING: The 'DateTimePicker' and all the sub components aren't ready to be used in Apps. Code can (will) change outside the release process until it's ready.",
		);

		const selectedDate = props.selection.date;
		const selectedTime = props.selection.time;

		const initialCalendarDate = selectedDate === undefined ? new Date() : selectedDate;

		this.state = {
			isDateTimeView: true,
			calendar: {
				monthIndex: getMonth(initialCalendarDate),
				year: getYear(initialCalendarDate),
			},
			selectedDate,
			selectedTime,
		};

		this.setDateTimeView = this.setView.bind(this, true);
		this.setMonthYearView = this.setView.bind(this, false);
		this.onSelectCalendarMonth = this.onSelectCalendarMonth.bind(this);
		this.onSelectCalendarYear = this.onSelectCalendarYear.bind(this);
		this.onSelectCalendarMonthYear = this.onSelectCalendarMonthYear.bind(this);
		this.onSelectDate = this.onSelectDate.bind(this);
		this.onSelectTime = this.onSelectTime.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		const selectionPropNotUpdated = this.props.selection === nextProps.selection;

		if (selectionPropNotUpdated) {
			return;
		}

		const newSelectedDate = nextProps.selection.date;
		const newSelectedTime = nextProps.selection.time;
		const needToUpdateDate = newSelectedDate !== this.state.selectedDate;
		const needToUpdateTime = newSelectedTime !== this.state.selectedTime;
		const noNeedToUpdateState = !needToUpdateDate && !needToUpdateTime;

		if (noNeedToUpdateState) {
			return;
		}

		const needToUpdateCalendar = needToUpdateDate && newSelectedDate !== undefined;

		const newState = {};

		if (needToUpdateDate) {
			newState.selectedDate = newSelectedDate;
		}

		if (needToUpdateTime) {
			newState.selectedTime = newSelectedTime;
		}

		if (needToUpdateCalendar) {
			newState.calendar = {
				monthIndex: getMonth(newSelectedDate),
				year: getYear(newSelectedDate),
			};
		}

		this.setState(newState);
	}

	onSelectDate(selectedDate) {
		this.setState({ selectedDate }, () => {
			this.trySubmit();
		});
	}

	onSelectTime(selectedTime) {
		this.setState({ selectedTime }, () => {
			this.trySubmit();
		});
	}

	onSelectCalendarMonthYear(newCalendar) {
		this.setState(previousState => ({
			calendar: {
				...previousState.calendar,
				...newCalendar,
			},
		}));
	}

	onSelectCalendarMonth(monthIndex) {
		this.onSelectCalendarMonthYear({ monthIndex });
	}

	onSelectCalendarYear(year) {
		this.onSelectCalendarMonthYear({ year });
	}

	setView(isDateTimeView) {
		this.setState({ isDateTimeView });
	}

	trySubmit() {
		if (this.state.selectedDate !== undefined && this.state.selectedTime !== undefined) {
			this.props.onSubmit({
				date: this.state.selectedDate,
				time: this.state.selectedTime,
			});
		}
	}

	render() {
		let viewElement;

		if (this.state.isDateTimeView) {
			viewElement = (
				<DateTimeView
					onClickTitle={this.setMonthYearView}
					calendar={this.state.calendar}
					onSelectMonthYear={this.onSelectCalendarMonthYear}
					onSelectDate={this.onSelectDate}
					selectedDate={this.state.selectedDate}
					onSelectTime={this.onSelectTime}
					selectedTime={this.state.selectedTime}
				/>
			);
		} else {
			viewElement = (
				<MonthYearView
					onClickBack={this.setDateTimeView}
					selectedMonthIndex={this.state.calendar.monthIndex}
					selectedYear={this.state.calendar.year}
					onSelectMonth={this.onSelectCalendarMonth}
					onSelectYear={this.onSelectCalendarYear}
				/>
			);
		}

		return <div className={theme.container}>{viewElement}</div>;
	}
}

DateTimePicker.propTypes = {
	selection: PropTypes.shape({
		date: PropTypes.instanceOf(Date),
		time: PropTypes.number,
	}),
	onSubmit: PropTypes.func.isRequired,
};

DateTimePicker.defaultProps = {
	selection: {},
};

export default DateTimePicker;
