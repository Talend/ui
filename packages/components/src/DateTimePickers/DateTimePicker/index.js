import DateTimePicker from './DateTimePicker.component';

export default DateTimePicker;
