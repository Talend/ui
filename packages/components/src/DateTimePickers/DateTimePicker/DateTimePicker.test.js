import React from 'react';
import { shallow } from 'enzyme';

import { mockDate, restoreDate } from '../shared/utils/test/dateMocking';

import DateTimePicker from './DateTimePicker.component';
import DateTimeView from '../views/DateTimeView';
import MonthYearView from '../views/MonthYearView';

describe('DateTimePicker', () => {
	it('should render', () => {
		mockDate(new Date(2018, 5, 12));

		const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

		expect(wrapper.getElement()).toMatchSnapshot();

		restoreDate();
	});

	describe('view switching', () => {
		it('should render with the DateTimeView based on state', () => {
			const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

			wrapper.setState({
				isDateTimeView: true,
			});

			expect(wrapper.find(DateTimeView).exists()).toBe(true);
			expect(wrapper.find(MonthYearView).exists()).toBe(false);
		});

		it('should render with the MonthYearView on state', () => {
			const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

			wrapper.setState({
				isDateTimeView: false,
			});

			expect(wrapper.find(DateTimeView).exists()).toBe(false);
			expect(wrapper.find(MonthYearView).exists()).toBe(true);
		});

		it('should switch state to MonthYearView when header title of DateTimeView is clicked', () => {
			const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

			wrapper.setState({
				isDateTimeView: true,
			});

			const clickTitleHandler = wrapper.find(DateTimeView).prop('onClickTitle');
			clickTitleHandler();
			expect(wrapper.state('isDateTimeView')).toBe(false);
		});

		it('should switch state to DateTimeView when header back action of MonthYearView is clicked', () => {
			const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

			wrapper.setState({
				isDateTimeView: false,
			});

			const clickBackHandler = wrapper.find(MonthYearView).prop('onClickBack');
			clickBackHandler();
			expect(wrapper.state('isDateTimeView')).toBe(true);
		});
	});

	it('should have the last "date" data between props and selection in state', () => {
		const d1 = new Date(2018, 2, 5);
		const d2 = new Date(2019, 11, 21);
		const d3 = new Date(2015, 6, 27);
		const wrapper = shallow(
			<DateTimePicker
				selection={{
					date: d1,
				}}
				onSubmit={() => {}}
			/>,
		);

		wrapper.setState({
			isDateTimeView: true,
		});

		const dateTimeView = wrapper.find(DateTimeView);
		expect(wrapper.state('selectedDate')).toBe(d1);

		dateTimeView.prop('onSelectDate')(d2);
		expect(wrapper.state('selectedDate')).toBe(d2);

		wrapper.setProps({
			selection: {
				date: d3,
			},
		});
		expect(wrapper.state('selectedDate')).toBe(d3);
	});

	it('should have the last "time" data between props and selection in state', () => {
		const t1 = 810;
		const t2 = 950;
		const t3 = 1270;
		const wrapper = shallow(
			<DateTimePicker
				selection={{
					time: t1,
				}}
				onSubmit={() => {}}
			/>,
		);

		wrapper.setState({
			isDateTimeView: true,
		});

		const dateTimeView = wrapper.find(DateTimeView);
		expect(wrapper.state('selectedTime')).toBe(t1);

		dateTimeView.prop('onSelectTime')(t2);
		expect(wrapper.state('selectedTime')).toBe(t2);

		wrapper.setProps({
			selection: {
				time: t3,
			},
		});
		expect(wrapper.state('selectedTime')).toBe(t3);
	});

	it('should not update state if props selection wrapper has not changed', () => {
		const initDate = new Date(2018, 0, 1);
		const initTime = 1000;
		const selection = {
			date: initDate,
			time: initTime,
		};

		const wrapper = shallow(<DateTimePicker selection={selection} onSubmit={() => {}} />);

		selection.date = new Date(2020, 0, 1);
		selection.time = 50;

		wrapper.setProps({
			selection,
		});

		expect(wrapper.state('selectedDate')).toBe(initDate);
		expect(wrapper.state('selectedTime')).toBe(initTime);
	});

	it('should auto submit when a date or time is selected and both are selected at this time', () => {
		const date = new Date(2018, 1, 1);
		const time = 1000;
		const onSubmit = jest.fn();
		const wrapper = shallow(
			<DateTimePicker
				selection={{
					date,
				}}
				onSubmit={onSubmit}
			/>,
		);

		wrapper.setState({
			isDateTimeView: true,
		});

		// TODO: Switch with the future internal submit button directly
		const dateTimeView = wrapper.find(DateTimeView);
		dateTimeView.prop('onSelectTime')(time);

		expect(onSubmit).toHaveBeenCalledWith({
			date,
			time,
		});
	});

	it('should not submit if one of date and time is not selected', () => {
		const time = 1000;
		const onSubmit = jest.fn();
		const wrapper = shallow(<DateTimePicker onSubmit={onSubmit} />);

		wrapper.setState({
			isDateTimeView: true,
		});

		const dateTimeView = wrapper.find(DateTimeView);
		dateTimeView.prop('onSelectTime')(time);

		expect(onSubmit).not.toHaveBeenCalled();
	});

	describe('calendar', () => {
		it('should at initialization define the calendar displayed based on current date when no selection props given', () => {
			mockDate(new Date(2016, 4, 12));

			const wrapper = shallow(<DateTimePicker onSubmit={() => {}} />);

			wrapper.setState({
				isDateTimeView: true,
			});

			const dateTimeView = wrapper.find(DateTimeView);
			expect(dateTimeView.prop('calendar')).toEqual({
				monthIndex: 4,
				year: 2016,
			});

			restoreDate();
		});

		it('should at initialization define the calendar displayed based on date selection prop when given', () => {
			const wrapper = shallow(
				<DateTimePicker
					selection={{
						date: new Date(2013, 0, 15),
					}}
					onSubmit={() => {}}
				/>,
			);

			wrapper.setState({
				isDateTimeView: true,
			});

			const dateTimeView = wrapper.find(DateTimeView);
			expect(dateTimeView.prop('calendar')).toEqual({
				monthIndex: 0,
				year: 2013,
			});
		});

		it('should update the calendar displayed based on date selection prop update', () => {
			const wrapper = shallow(
				<DateTimePicker
					selection={{
						date: new Date(2013, 0, 15),
					}}
					onSubmit={() => {}}
				/>,
			);

			wrapper.setState({
				isDateTimeView: true,
			});

			const dateTimeViewBefore = wrapper.find(DateTimeView);
			expect(dateTimeViewBefore.prop('calendar')).toEqual({
				monthIndex: 0,
				year: 2013,
			});

			wrapper.setProps({
				selection: {
					date: new Date(2015, 9, 4),
				},
			});

			const dateTimeViewAfter = wrapper.find(DateTimeView);
			expect(dateTimeViewAfter.prop('calendar')).toEqual({
				monthIndex: 9,
				year: 2015,
			});
		});

		it('should keep actual calendar displayed if date selection prop update to undefined', () => {
			const wrapper = shallow(
				<DateTimePicker
					selection={{
						date: new Date(2013, 0, 15),
					}}
					onSubmit={() => {}}
				/>,
			);

			wrapper.setState({
				isDateTimeView: true,
			});

			const dateTimeViewBefore = wrapper.find(DateTimeView);
			expect(dateTimeViewBefore.prop('calendar')).toEqual({
				monthIndex: 0,
				year: 2013,
			});

			wrapper.setProps({
				selection: {
					date: undefined,
				},
			});

			const dateTimeViewAfter = wrapper.find(DateTimeView);
			expect(dateTimeViewAfter.prop('calendar')).toEqual({
				monthIndex: 0,
				year: 2013,
			});
		});
	});
});
