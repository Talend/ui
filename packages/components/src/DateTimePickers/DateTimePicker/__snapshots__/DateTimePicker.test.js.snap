// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker should render 1`] = `
<div
  aria-label="Date picker"
  className="theme-container"
  tabIndex={0}
>
  <Translate(DateTimeView)
    allowFocus={true}
    calendar={
      Object {
        "monthIndex": 5,
        "year": 2018,
      }
    }
    onSelectDate={[Function]}
    onSelectMonthYear={[Function]}
    onSelectTime={[Function]}
    onTitleClick={[Function]}
    selectedDate={undefined}
    selectedTime={undefined}
    useSeconds={undefined}
    useTime={undefined}
  />
</div>
`;
